pilot/pkg/proxy/envoy/v2/ep_filters.go

```diff
@@ -35,118 +35,123 @@ func EndpointsByNetworkFilter(endpoints []endpoint.LocalityLbEndpoints, conn *Xd
 			}
 		}
 
-		if !foundRemote {
-			// This LocalityLbEndpoints has no remote endpoints so just
-			// add it to the result
-			filtered = append(filtered, ep)
-		} else {
-			if len(onlyLocalLbEndpoints) > 0 {
-				// This LocalityLbEndpoints has remote endpoint so add to the result
-				// a new one that holds only local endpoints
-				newEp := endpoint.LocalityLbEndpoints{
-					Locality:    ep.Locality,
-					LbEndpoints: onlyLocalLbEndpoints,
-					LoadBalancingWeight: &types.UInt32Value{
-						Value: uint32(len(onlyLocalLbEndpoints)),
-					},
-					Priority: ep.Priority,
-				}
-				filtered = append(filtered, newEp)
-			}
-		}
-	}
-
-	// If there is no MeshNetworks configuration, we don't have gateways information
-	// so just return the filtered results (with no remote endpoints)
-	if env.MeshNetworks == nil {
-		return filtered
-	}
-
-	// Iterate over all networks that have the cluster endpoint (weight>0) and
-	// for each one of those add a new endpoint that points to the network's
-	// gateway with the relevant weight
-	for n, w := range remoteEps {
-		networkConf, found := env.MeshNetworks.Networks[n]
-		if !found {
-			continue
-		}
-		gws := networkConf.Gateways
-		if len(gws) == 0 {
+		// If there is no MeshNetworks configuration, we don't have gateways information
+		// so just keep the endpoint with the local ones
+		if env.MeshNetworks == nil {
+			newEp := createLocalityLbEndpoints(&ep, lbEndpoints)
+			filtered = append(filtered, *newEp)
 			continue
 		}
 
-		// There may be multiple gateways for the network. Add an LbEndpoint for
-		// each one of them
-		gwEndpoints := []endpoint.LbEndpoint{}
-		for _, gw := range gws {
-			// If the gateway address is DNS we can't add it to the endpoints
-			if gwIP := net.ParseIP(gw.GetAddress()); gwIP != nil {
-				addr := util.BuildAddress(gw.GetAddress(), gw.Port)
-				gwEp := endpoint.LbEndpoint{
-					Endpoint: &endpoint.Endpoint{
-						Address: &addr,
-					},
-				}
-				gwEndpoints = append(gwEndpoints, gwEp)
+		// Add endpoints to remote networks' gateways
+
+		// Iterate over all networks that have the cluster endpoint (weight>0) and
+		// for each one of those add a new endpoint that points to the network's
+		// gateway with the relevant weight
+		for n, w := range remoteEps {
+			networkConf, found := env.MeshNetworks.Networks[n]
+			if !found {
+				continue
+			}
+			gws := networkConf.Gateways
+			if len(gws) == 0 {
+				continue
 			}
 
-			if gw.GetRegistryServiceName() != "" {
+			// There may be multiple gateways for the network. Add an LbEndpoint for
+			// each one of them
+			for _, gw := range gws {
+				var gwEp *endpoint.LbEndpoint
 				//TODO add support for getting the gateway address from the service registry
-			}
-		}
 
-		// No endpoints for this gatway, don't add it
-		if len(gwEndpoints) == 0 {
-			continue
-		}
+				// If the gateway address is DNS we can't add it to the endpoints
+				if gwIP := net.ParseIP(gw.GetAddress()); gwIP != nil {
```

Comment:
do you want to try to resolve this during config load, so that you can send IPs to envoys?

Reply:
I can and will (thanks for the idea) but this is not perfect solution as the IP may change over time.